File: bindings/R/FortMyrmidon/R/Ant.R
#' @name fmCAnt
#' @title A reference class that represent a non-modifiable Ant
#' @description This is a wrapper for the C++ class
#'     \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon::CAnt}{fort::myrmidon::CAnt}
#'
#' @field cIdentifications
#'     \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon::CAnt/fort::myrmidon::CAnt::CIdentifications}{fort::myrmidon::CAnt::CIdentifications}
#' @field antID
#'     \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon::CAnt/fort::myrmidon::CAnt::AntID}{fort::myrmidon::CAnt::AntID}
#' @field formattedID
#'     \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon::CAnt/fort::myrmidon::CAnt::FormattedID}{fort::myrmidon::CAnt::FormattedID}
#' @field displayColor
#'     \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon::CAnt/fort::myrmidon::CAnt::DisplayColor}{fort::myrmidon::CAnt::DisplayColor}
#' @field displayStatus
#'     \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon::CAnt/fort::myrmidon::CAnt::DisplayStatus}{fort::myrmidon::CAnt::DisplayStatus}
#' @field getValue
#'     \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon::CAnt/fort::myrmidon::CAnt::GetValue}{fort::myrmidon::CAnt::GetValue}
#' @field capsules wrapper for
#'     \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon::CAnt/fort::myrmidon::CAnt::Capsules}{fort::myrmidon::CAnt::Capsules}. It
#'     returns a list with two elements, the 'capsules' element, a
#'     list of \linkS4class{fmCapsule} and 'shapeTypes' their
#'     corresponding shape types.
NULL


#' @name fmAnt
#' @title A reference class that represent an Ant
#' @description This is a wrapper for the C++ class
#'     \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon::Ant}{fort::myrmidon::Ant}
#'
#' @field cIdentifications
#'     \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon::Ant/fort::myrmidon::Ant::CIdentifications}{fort::myrmidon::Ant::CIdentifications}
#' @field identifications
#'     \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon::Ant/fort::myrmidon::Ant::Identifications}{fort::myrmidon::Ant::Identifications}
#' @field antID
#'     \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon::Ant/fort::myrmidon::Ant::AntID}{fort::myrmidon::Ant::AntID}
#' @field formattedID
#'     \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon::Ant/fort::myrmidon::Ant::FormattedID}{fort::myrmidon::Ant::FormattedID}
#' @field displayColor
#'     \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon::Ant/fort::myrmidon::Ant::DisplayColor}{fort::myrmidon::Ant::DisplayColor}
#' @field setDisplayColor
#'     \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon::Ant/fort::myrmidon::Ant::SetDisplayColor}{fort::myrmidon::Ant::SetDisplayColor}
#' @field displayStatus
#'     \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon::Ant/fort::myrmidon::Ant::DisplayStatus}{fort::myrmidon::Ant::DisplayStatus}
#' @field getValue
#'     \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon::Ant/fort::myrmidon::Ant::GetValue}{fort::myrmidon::Ant::GetValue}
#' @field deleteValue
#'     \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon::Ant/fort::myrmidon::Ant::DeleteValue}{fort::myrmidon::Ant::DeleteValue}
#' @field setValue
#'     \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon::Ant/fort::myrmidon::Ant::SetValue}{fort::myrmidon::Ant::SetValue}
#' @field addCapsule
#'     \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon::Ant/fort::myrmidon::Ant::AddCapsule}{fort::myrmidon::Ant::AddCapsule}
#' @field capsules wrapper for
#'     \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon::Ant/fort::myrmidon::Ant::Capsules}{fort::myrmidon::Ant::Capsules}. It
#'     returns a list with two elements, the 'capsules' element, a
#'     list of \linkS4class{fmCapsule} and 'shapeTypes' their
#'     corresponding shape types.
#' @field deleteCapsule
#'     \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon::Ant/fort::myrmidon::Ant::DeleteCapsule}{fort::myrmidon::Ant::DeleteCapsule}
#' @field clearCapsules
#'     \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon::Ant/fort::myrmidon::Ant::ClearCapsules}{fort::myrmidon::Ant::ClearCapsules}
#' @field setDisplayStatus
#'     \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon::Ant/fort::myrmidon::Ant::SetDisplayStatus}{fort::myrmidon::Ant::SetDisplayStatus}
NULL

File: bindings/R/FortMyrmidon/src/ant.cpp
//#include "identification.h"

#include "ant.h"

#include "identification.h"
#include "color.h"
#include "ant_static_value.h"
#include "time.h"
#include "shapes.h"

namespace Rcpp {

template <> SEXP wrap(const fort::myrmidon::TypedCapsuleList & l );

}

#include "Rcpp.h"

namespace Rcpp {

template <> SEXP wrap(const fort::myrmidon::TypedCapsuleList & l ) {
	IntegerVector types;
	NumericVector c1X,c1Y,c1Radius,c2X,c2Y,c2Radius;
	List objects;

	for ( const auto & [shapeTypeID,c] : l ) {
		types.push_back(shapeTypeID);
		c1X.push_back(c.C1().x());
		c1Y.push_back(c.C1().y());
		c1Radius.push_back(c.R1());
		c2X.push_back(c.C2().x());
		c2Y.push_back(c.C2().y());
		c2Radius.push_back(c.R2());
		objects.push_back(c);
	}

	List res;
	res["summary"] = DataFrame::create(_["typeID"] = types,
	                                   _["c1.x"] = c1X,
	                                   _["c1.y"] = c1Y,
	                                   _["c1.radius"] = c1Radius,
	                                   _["c2.x"] = c2X,
	                                   _["c2.y"] = c2Y,
	                                   _["c2.radius"] = c2Radius);
	res["objects"] = objects;
	return res;
}

}


void fmAnt_show(const fort::myrmidon::Ant * ant) {
	Rcpp::Rcout << "fmAnt " << ant->FormattedID() << "\n";
}

void fmCAnt_show(const fort::myrmidon::CAnt * ant) {
	Rcpp::Rcout << "fmCAnt " << ant->FormattedID() << "\n";
}


RCPP_MODULE(ant) {
	Rcpp::class_<fort::myrmidon::CAnt>("fmCAnt")
		.const_method("show",&fmCAnt_show)
		.const_method("identifiedAt",&fort::myrmidon::CAnt::IdentifiedAt)
		.const_method("cIdentifications",&fort::myrmidon::CAnt::CIdentifications)
		.const_method("antID",&fort::myrmidon::CAnt::AntID)
		.const_method("formattedID",&fort::myrmidon::CAnt::FormattedID)
		.const_method("displayColor",&fort::myrmidon::CAnt::DisplayColor)
		.const_method("displayStatus",&fort::myrmidon::CAnt::DisplayStatus)
		.const_method("getValue",&fort::myrmidon::CAnt::GetValue)
		.const_method("capsules",&fort::myrmidon::CAnt::Capsules)
		;

	Rcpp::class_<fort::myrmidon::Ant>("fmAnt")
		.const_method("show",&fmAnt_show)
		.const_method("identifiedAt",&fort::myrmidon::Ant::IdentifiedAt)
		.const_method("cIdentifications",&fort::myrmidon::Ant::CIdentifications)
		.method("identifications",&fort::myrmidon::Ant::Identifications)
		.const_method("antID",&fort::myrmidon::Ant::AntID)
		.const_method("formattedID",&fort::myrmidon::Ant::FormattedID)
		.const_method("displayColor",&fort::myrmidon::Ant::DisplayColor)
		.method("setDisplayColor",&fort::myrmidon::Ant::SetDisplayColor)
		.const_method("displayStatus",&fort::myrmidon::Ant::DisplayStatus)
		.method("setDisplayStatus",&fort::myrmidon::Ant::SetDisplayStatus)
		.const_method("getValue",&fort::myrmidon::Ant::GetValue)
		.method("setValue",&fort::myrmidon::Ant::SetValue)
		.method("deleteValue",&fort::myrmidon::Ant::DeleteValue)
		.method("addCapsule",&fort::myrmidon::Ant::AddCapsule)
		.const_method("capsules",&fort::myrmidon::Ant::Capsules)
		.method("deleteCapsule",&fort::myrmidon::Ant::DeleteCapsule)
		.method("clearCapsules",&fort::myrmidon::Ant::ClearCapsules)
		;

}
